### DISCUSSION TITLE
Porous Flow Peaceman Borehole: injecting cold CO2 in a doublet

### DISCUSSION THREAD

**Discussion by fparisio** [idaholab/moose#21320] [Q&A Modules: General] [answered]
I have built a model that reproduces a doublet system where cold CO2 is injected into an aquifer from a first wellbore, and simultaneously fluid is produced at another wellbore. When the CO2 plume reaches the production wellbore, the produced fluid is a mixture of water and CO2.

The initial temperature is 90 °C and CO2 is injected at 60 °C. I am employing the PorousFlow2PhasePS material in non-isothermal conditions. Both injection and production are represented with PorousFlowPeacemanBorehole sinks/sources. The initial pressure is 27 MPa, CO2 is injected at 37 MPa while the fluid is produced at 24 MPa. The reservoir is 100 m thick and gravity is neglected.

I have observed that the pressure at the injection point is decreasing with time: why is this occurring and why is the pressure not constant and identical to the value specified in the PorousFlowPeacemanBorehole command? I am having a hard time understanding the reason behind this. 

To inject CO2, I am assigning "sgas" as the variable in the injection PorousFlowPeacemanBorehole. If I change it to water injection "pwater", then the pressure is kept constant and at 37 MPa as specified by the PorousFlowPeacemanBorehole. I suspect it has to do with the relative permeability, which is likely influencing the mobility term in the PorousFlowPeacemanBorehole model, but I do not understand why P_wellbore is changing.

I have attached the files of the simulation, along with a graph showing the evolution of CO2 and water pressure and CO2 saturation at the injection well.

Any suggestion is highly appreciated!

![injection_well](https://user-images.githubusercontent.com/92143809/174101618-616d5723-a9c1-454f-b718-672ccfaab38e.png)

[CO2-Injection.tar.gz](https://github.com/idaholab/moose/files/8920112/CO2-Injection.tar.gz)

**Comment by WilkAndy** [comment, marked as answer]
Hi @fparisio , i haven't run your simulation yet, sorry, but here are a few thoughts:
- the porepressure "at" a peaceman borehole is often not exactly what you specify in the input file.  This is by design.  Here is an explanation....  The Peaceman formulation works by using the analytical solution of the fluid pressure in an element: some further remarks are at https://mooseframework.inl.gov/modules/porous_flow/sinks.html .  More specifically, if your bottomhole pressure is P within an element (you specify P in the `PorousFlowPeacemanBorehole`), then MOOSE's finite-element solution at the nodes of that element will generally not be P.  Your postprocessor(s) will be equal (or related) to MOOSE's solution directly, not directly to your bottomhole pressure.  The discrepancy you see will be problem dependent.  For example, if the finite element containing your Peaceman borehole is huge, then the porepressure at its nodes will be quite different from bottomhole pressure, while if your finite element is small, the element's nodal pressures will be similar to the bottomhole pressure.  As the finite element size limits to the size of the borehole, the nodal pressures limit to the bottomhole pressure.
- the reason why the pressure is changing could be the mobility, since physically it's likely to be getting easier and easier to pump in CO2 (a CO2-saturated reservoir usually has higher mobility).  However, i also suspect that it's a temperature effect.  As you cool the reservoir, the fluids will contract and the pressure will decrease - sometimes quite significantly.  i suggest running a simulation at 90degC and exploring the difference.
- Now a question from me: on your figure, why does Pwater = PCO2 when gas saturation = 1.  Usually capillarity would mean Pwater << PCO2 at that saturation.
a

**Reply by fparisio** [reply]
Hi @WilkAndy, thank you for the prompt and detailed replies! Below are some comments on your points:

- I understand a bit better the principles of the ```PorousFlowPeacemanBorehole```, especially this element-size dependency. I have run a model with smaller elements and I saw that the pressure is decreasing earlier than compared with larger elements;
- Thanks for the hint, I have tried an isothermal case and the changes in pressure compared to cold injection are very small (but non exactly 0, see the plot). Concerning mobility, I have tried a case with constant relative permeability but it does not converge. Perhaps it is because of null capillary force (my guess is that constant pc and rel_perm make it difficult for the PS formulation to calculate sgas);
- Capillarity is null because I fixed capillary pressure pc=0 on purpose, I wanted to isolate the effects for now. But your hint is correct, I have tried with the capillary pressure and the results show that indeed higher retention properties (VanGenuchten with alpha=8e-7 and m=0.6) keep the gas pressure closer to the prescribed values from the ```PorousFlowPeacemanBorehole```. One question: is there a way to make the retention properties temperature-dependent? For example, through temperature-dependent surface tension?
- I have another question on the usage of ```PorousFlowPeacemanBorehole``` in a true 2D setting. I have run an example of that but I do not fully understand what would be the meaning of the out-of-plane size of the wellbore? A 2D model is usually assumed as unit thickness in the out-of-plane direction, so what would be for example a ```PorousFlowPeacemanBorehole``` with 2 points, one of which with a z coordinate (assuming 2D is in x,y) is 100 m? I have run some tests and observed that the injected mass flow rate reaches an asymptotic limit with the increasing out-of-plane size of the ```PorousFlowPeacemanBorehole```. I have tried different sizes from 1m to 3000m and found that at approximately 500m out-of-plane well length, the mass flow rate was 99.2% the one of 3000m wellbore size. I am not sure about its physical interpretation, or if it is even correct to use ```PorousFlowPeacemanBorehole``` in a 2D case.

The figure below shows the results of the different trials. The light grey plots indicate the base case, for comparison. 

![injection_well](https://user-images.githubusercontent.com/92143809/174280355-026d5253-0fab-4743-82b0-8e3d66cbb6de.png)

**Reply by WilkAndy** [reply]
Hi @fparisio 

- sorry, it's not possible to make the retention temperature-dependent at the moment.  It would fit into the PorousFlow framework well, but we just haven't coded such a thing.  Do you have experimental data for that, by the way?
- regarding the out-of-plane question: yes, you can use the Peaceman in that situation, and usually the out-of-plane length is assumed to be 1m (or, equivalently, everything is calculated per unit thickness), both in the conceptual model, and in the `PorousFlowPeacemanBorehole`.  (You may assume otherwise if you like, but it's unusual).  Often people supply just one point to the PeacemanBorehole (with the "z" direction = zero, for instance) along with `line_length=1` and `line_direction='0 0 1'`.
- I'm not sure whether the following suits your problem or not....   You might like to consider another boundary condition for the borehole injection.  For instance, if you're fixing the injection rate, or your injection regime is a known function of porepressure in the rock, then [PorousFlowPolyLineSink](https://mooseframework.inl.gov/source/dirackernels/PorousFlowPolyLineSink.html) might be better.  Or, if you're fixing the porepressure, then a DirichletBC might be OK (but remember a DirichletBC is fixing porepressure by injecting/removing a fluid, so you have to be careful which fluid component (Variable) you're applying the DirichletBC to).

a

**Reply by fparisio** [reply]
Hi @WilkAndy, again, thanks a lot for the prompt reply! Here are some comments:

- Good question, but I think that for CO2-H2O mixture there is not much data. One simplistic approach is to scale entry pressure $p_a$ with surface tension $s$ as 
$$p_a = p_{a0} \frac{s}{s_0}$$
where surface tension $s_0$ and entry pressure $p_{a0}$ are measured at reference conditions. Then, one could assume the variations as a function of temperature and pressure through $s(p, T)$, assuming experimental results (often in the form of MD simulations) for the latter are available as in, e.g., Nielsen et al. (2012). For higher temperature applications, there was an interesting work at the last EGU by Shiga et al. (2022), but I believe the journal paper is not out yet. In any case, it is definetely a relevant and underexplored topic.

- Thank you for the suggestion on how to use the ```PorousFlowPeacemanBorehole``` in a 2D setting.

- I thought about using DirichletBC and fixing pressure, but I am not sure how to compute the extracted/injected masses at a node from the ```PorousFlowDarcyVelocityComponent``` module in ```AuxKernels```. Is there a way to compute nodal flow mass rates (volumetric or mass) once fixed a DirichletBC?

References:

Nielsen, L.C., Bourg, I.C. and Sposito, G., 2012. Predicting CO2–water interfacial tension under pressure and temperature conditions of geologic CO2 storage. Geochimica et Cosmochimica Acta, 81, pp.28-38.

Shiga, M., Morishita, T. and Sorai, M., 2022. Interfacial Tension of CO2-Water Under Conditions for High-Temperature Geothermal Systems: Prediction and Investigation by Molecular Dynamics Simulation (No. EGU22-6837). Copernicus Meetings.

**Reply by WilkAndy** [reply]
Thanks for those refs, @fparisio .   One ugly way that you could calculate mass flowrates through a "Dirichlet borehole" is to construct a sideset around the borehole, then use a [PorousFlowOutflowBC](https://mooseframework.inl.gov/source/bcs/PorousFlowOutflowBC.html) with a `save_in` to record the mass-flow through each node on that sideset (and example of `save_in` is found in test/tests/sinks/injection_production_eg_outflowBC.i). Then use another almost-identical `PorousFlowOutflowBC`, but with `multiplier = -1`.  The result is that the residual gets a contribution of +flow-flow (=0) so it is unaffected, but you record the flow in the `save_in` AuxVariable.  This is quite ugly!

a

**Reply by fparisio** [reply]
Hi @WilkAndy and thank you for the explanation. If I understood correctly, for example for the injection point, I will add two ```PorousFlowOutflowBC``` to cancel out the contribution, but record one of them to save it and be able to post-process it as a mass flow rate, correct? I have added what I think is a working example of the injection of CO2 (no capillarity) with a ```DirichletBC``` and with ```PorousFlowOutflowBC``` for the mass flow rate. Did I make a correct interpretation?
```
[BCs]
  [injection_dirichlet_saturation]
    type = DirichletBC
    variable = sgas
    boundary = injection_node
    value = 1
  []
  [injection_dirichlet_pressure]
    type = DirichletBC
    variable = pwater
    boundary = injection_node
    value = 37e6
  []
  [inj_co2_component1]
    type = PorousFlowOutflowBC
    boundary = injection_node
    variable = sgas
    mass_fraction_component = 1
    save_in = co2_kg_per_s
    multiplier = 1
  []
  [inj_co2_component1_mirror]
    type = PorousFlowOutflowBC
    boundary = injection_node
    variable = sgas
    mass_fraction_component = 1
    multiplier = -1
  []
[]
```
and then add it to the postprocessor as
```
  [co2_kg_per_s]
    type = NodalSum
    boundary = injection_node
    variable = co2_kg_per_s
  []
```

**Reply by WilkAndy** [reply]
@fparisio , you got the concept correct, yet.  There are some caveats:

- the `PorousFlowOutflowBC`s need to be applied to a boundary that contains element edges, not just a single node.  So you should create a boundary in your mesh that surrounds the injection node.  The normal to the boundary should be oriented "outward" (if you get this wrong then you'll just record a negative co2_kg_per_s, which is no big deal).
- the `variable` and `mass_fraction_component` must be set correctly.  i can't tell from the above snippet whether you've done this correctly, but you probably have.  If the sgas `variable` is looking after `fluid_component = 1` in the Kernels, and the mass-fraction of fluid_component = 1 is all in phase = 1, and the mass-fraction of fluid_component = 0 is all in phase = 0 then you're currently recording phase 1 (probably gas), fluid_component 1 (probably CO2) coming through the boundary.

I'm not so sure of your `DirichletBC`s.  You need to look at your `Kernels` to see which `fluid_component` each variable (sgas and pwater) are looking after.  Also, your `PorousFlowMassFraction` `Materials` to see which phase those fluid components correspond to.  The important thing is to physically interpret the `DirichletBC`s as injecting/removing fluid components in order to satisfy the DirichletBC.  For instance, at the moment the `injection_dirichlet_saturation` might be fixing `sgas = 1` by injecting fluid_component = 1 into the system (which is probably in CO2, and is probably in phase 1, which is probably gas), and the  `injection_dirichlet_pressure` might be fixing `pwater = 37e6` by injecting fluid_component = 0 into the system (which is probably H2O, and is probably in phase 0, which is probably liquid).  The "probably" in the previous sentence can become "definitely" when you explore your `Kernels` and `Materials`.

**Reply by WilkAndy** [reply]
@fparisio , when i say a "boundary" in the above, i mean a sideset in the mesh.  Sidesets needn't be on the actual boundary of the mesh: they can be internal to the mesh, as yours will be.  Sidesets are just a collection of element sides (for 3D elements) or edges (for 2D elements)

**Reply by fparisio** [reply]
Hi @WilkAndy, thank you for clarifying the meaning of sideset. 

Concerning the injection problem, the way I set up the example is a consequence of my interpretation of the ```PorousFlow2PhasePS```: if the CO2 is represented by its saturation (sgas), is there a better way to simulate CO2 injection at constant pressure, i.e., as ```DirichletBC```? In the example that I brought it works because with null capillarity pgas==pwater, but could one fix pgas instead at the boundary?

**Reply by WilkAndy** [reply]
Possibly, you would need to re-phrase your input file (note, i haven't checked exactly how your input file is constructed, so maybe it's perfect now).  Physically, your pumps are injecting CO2 at a fixed pressure.  So you want to fix Pgas with a DirichletBC, and you want the `Variable` Pgas to be looking after the CO2 component in the `Kernels`.  Eg, if component = 1 represents CO2, then you want component=1 in all the Pgas Kernels (and you want to make sure the mass-fractions are set to 1 and/or 0 appropriately elsewhere in your input file).   Since your pumps are not concerning H2O at all, you don't want any other DirichletBC.   One of the reasons for the Peaceman approach is that you don't have to worry about all this stuff!

**Reply by fparisio** [reply]
Thanks a lot, @WilkAndy. I also came to use the ```PorousFlowPeacemanBorehole``` for the same reasons. I appreciate your help and your comments allowed me to better understand how to simulate the problem in MOOSE.
